firebase-auth-claims extension does not run against the emulator

## Expected Behavior

The firebase-auth-claims sets custom claims using the authentication emulator.

## Actual Behavior

The firestore `user_claims` document is created in the emulator. I can see the `user_claims` object and add attributes. 

The extension attempts to update claims in the _cloud_ instance, not the local auth emulator.  I can validate this by using a uid from the cloud. When the local firestore emulator user_claims is written with a cloud uid, it updates the cloud ok - no error. 

When I attempt to use a local auth emulator id, I get " error: FirebaseAuthError: There is no user record corresponding to the provided identifier". This uid does not exist in the cloud. I understand the error - but not how to get the extension to use the auth emulator.  

My other functions work fine in the emulator, and clearly the extension is being triggered by a write to the firestore emulator instance. It's just not going against the local auth emulator.

I also get a "'firebase-admin' dependency is below version 9.3.0," It looks like the extension needs to be updated?

## Steps to Reproduce the Problem

What happened? How can we make the problem occur?

Install the extension using the cli
Add a doc to user_claims with the id of an existing user in the authentication emulator.  

## Specifications

- Extension name: firebase-auth-claims
- Version: 0.1.1

Also, the emulator crashes when the extension is invoked:
```
!  The Cloud Functions emulator requires the module "firebase-admin" to be installed. This package is in your package.json, but it's not available. You probably need to run "npm install" in your functions directory.
i  functions: Your functions could not be parsed due to an issue with your node_modules (see above)
!  Error: Cannot find module 'firebase-functions'
Require stack:
- C:\Users\dan\.cache\firebase\extensions\firebase\firestore-auth-claims@0.1.1\functions\lib\index.js
```

--- a/_emulator/extensions/firestore-auth-claims.env.local
+++ b/_emulator/extensions/firestore-auth-claims.env.local
@@ -1,0 +1,1 @@
+LOCATION=europe-west2

--- a/_emulator/firebase.json
+++ b/_emulator/firebase.json
@@ -1,7 +1,6 @@
 {
   "extensions": {
-    "firestore-record-user-acknowledgements": "../firestore-record-user-acknowledgements",
-    "firestore-bundle-server": "../firestore-bundle-server"
+    "firestore-auth-claims": "../firestore-auth-claims"
   },
   "storage": {
     "rules": "storage.rules"

--- a/_emulator/functions/.gitignore
+++ b/_emulator/functions/.gitignore
@@ -1,0 +1,9 @@
+# Compiled JavaScript files
+lib/**/*.js
+lib/**/*.js.map
+
+# TypeScript v1 declaration files
+typings/
+
+# Node.js dependency directory
+node_modules/

--- a/_emulator/functions/package.json
+++ b/_emulator/functions/package.json
@@ -1,0 +1,25 @@
+{
+  "name": "functions",
+  "scripts": {
+    "build": "tsc",
+    "build:watch": "tsc --watch",
+    "serve": "npm run build && firebase emulators:start --only functions",
+    "shell": "npm run build && firebase functions:shell",
+    "start": "npm run shell",
+    "deploy": "firebase deploy --only functions",
+    "logs": "firebase functions:log"
+  },
+  "engines": {
+    "node": "16"
+  },
+  "main": "lib/index.js",
+  "dependencies": {
+    "firebase-admin": "^11.4.1",
+    "firebase-functions": "^4.1.0",
+    "@types/express-serve-static-core": "4.17.30"
+  },
+  "devDependencies": {
+    "typescript": "^4.6.4"
+  },
+  "private": true
+}

--- a/_emulator/functions/src/index.ts
+++ b/_emulator/functions/src/index.ts
@@ -1,0 +1,8 @@
+import * as functions from "firebase-functions";
+
+export const findDocumentReferences = functions.https.onRequest(
+  (request, response) => {
+    /** Simply send back the data that we have posted */
+    response.send(["searchFunction/testing/functions-testing/example"]);
+  }
+);

--- a/_emulator/functions/tsconfig.json
+++ b/_emulator/functions/tsconfig.json
@@ -1,0 +1,13 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "noImplicitReturns": true,
+    "noUnusedLocals": true,
+    "outDir": "lib",
+    "sourceMap": true,
+    "strict": true,
+    "target": "es2017"
+  },
+  "compileOnSave": true,
+  "include": ["src"]
+}

--- a/firestore-auth-claims/functions/.gitignore
+++ b/firestore-auth-claims/functions/.gitignore
@@ -1,0 +1,9 @@
+# Compiled JavaScript files
+lib/**/*.js
+lib/**/*.js.map
+
+# TypeScript v1 declaration files
+typings/
+
+# Node.js dependency directory
+node_modules/

--- a/firestore-auth-claims/functions/package.json
+++ b/firestore-auth-claims/functions/package.json
@@ -4,21 +4,20 @@
     "lint": "eslint \"src/**/*\"",
     "build": "tsc",
     "generate-readme": "node ../../generate-experimental-readme.js firestore-auth-claims > ../README.md"
-
   },
   "main": "lib/index.js",
   "dependencies": {
-    "@types/node": "^12.19.3",
-    "@typescript-eslint/eslint-plugin": "^4.6.0",
-    "@typescript-eslint/parser": "^4.6.0",
-    "firebase-admin": "^8.6.0",
-    "firebase-functions": "^3.6.1"
+    "@types/node": "^18.14.6",
+    "@typescript-eslint/eslint-plugin": "^5.54.1",
+    "@typescript-eslint/parser": "^5.54.1",
+    "firebase-admin": "^11.5.0",
+    "firebase-functions": "^4.2.1"
   },
   "devDependencies": {
     "eslint": "^7.6.0",
     "eslint-plugin-import": "^2.22.0",
     "prettier": "^2.1.2",
-    "typescript": "^4.1.3"
+    "typescript": "^4.9.5"
   },
   "private": true
 }

--- a/firestore-auth-claims/functions/src/index.ts
+++ b/firestore-auth-claims/functions/src/index.ts
@@ -7,76 +7,80 @@ admin.initializeApp();
 const auth = admin.auth();
 
 const CLAIMS_FIELD: string | null = process.env.CLAIMS_FIELD || null;
+const CLAIMS_COLLECTION = process.env.CLAIMS_COLLECTION || "user_claims";
 
-exports.sync = functions.handler.firestore.document.onWrite(async (change) => {
-  const uid = change.after.id;
-  try {
-    // make sure the user exists (can be fetched) before trying to set claims
-    await auth.getUser(uid);
-  } catch (e) {
-    functions.logger.error(
-      `Unable to sync claims for user '${uid}', error:`,
-      e,
-      { uid }
-    );
-  }
+exports.sync = functions.firestore
+  .document(CLAIMS_COLLECTION)
+  .onWrite(async (change) => {
+    const uid = change.after.id;
+    try {
+      // make sure the user exists (can be fetched) before trying to set claims
+      await auth.getUser(uid);
+    } catch (e) {
+      functions.logger.error(
+        `Unable to sync claims for user '${uid}', error:`,
+        e,
+        { uid }
+      );
+    }
 
-  if (
-    !change.after.exists ||
-    (CLAIMS_FIELD && !change.after.get(CLAIMS_FIELD))
-  ) {
+    if (
+      !change.after.exists ||
+      (CLAIMS_FIELD && !change.after.get(CLAIMS_FIELD))
+    ) {
+      functions.logger.info(
+        `Claims for user '${uid}' were deleted, removing from Auth.`,
+        { uid }
+      );
+      return auth.setCustomUserClaims(uid, null);
+    }
+
+    const beforeData =
+      (CLAIMS_FIELD ? change.before.get(CLAIMS_FIELD) : change.before.data()) ||
+      {};
+    const data =
+      (CLAIMS_FIELD ? change.after.get(CLAIMS_FIELD) : change.after.data()) ||
+      {};
+    // don't write the _synced field to Auth
+    if (data._synced) {
+      delete data._synced;
+    }
+    if (beforeData._synced) {
+      delete beforeData._synced;
+    }
+
+    if (isDeepStrictEqual(beforeData, data)) {
+      // don't persist identical claims
+      return;
+    }
+
     functions.logger.info(
-      `Claims for user '${uid}' were deleted, removing from Auth.`,
+      `Updating claims for user '${uid}', setting keys ${Object.keys(data).join(
+        ", "
+      )}.`,
       { uid }
     );
-    return auth.setCustomUserClaims(uid, null);
-  }
+    if (typeof data !== "object") {
+      functions.logger.error(
+        `Invalid custom claims for user '${uid}'. Must be object, was ${JSON.stringify(
+          data
+        )}`,
+        { uid }
+      );
+      return;
+    }
+    await auth.setCustomUserClaims(uid, data);
 
-  const beforeData =
-    (CLAIMS_FIELD ? change.before.get(CLAIMS_FIELD) : change.before.data()) ||
-    {};
-  const data =
-    (CLAIMS_FIELD ? change.after.get(CLAIMS_FIELD) : change.after.data()) || {};
-  // don't write the _synced field to Auth
-  if (data._synced) {
-    delete data._synced;
-  }
-  if (beforeData._synced) {
-    delete beforeData._synced;
-  }
-
-  if (isDeepStrictEqual(beforeData, data)) {
-    // don't persist identical claims
-    return;
-  }
-
-  functions.logger.info(
-    `Updating claims for user '${uid}', setting keys ${Object.keys(data).join(
-      ", "
-    )}.`,
-    { uid }
-  );
-  if (typeof data !== "object") {
-    functions.logger.error(
-      `Invalid custom claims for user '${uid}'. Must be object, was ${JSON.stringify(
-        data
-      )}`,
+    const fpath = ["_synced"];
+    if (CLAIMS_FIELD) {
+      fpath.unshift(CLAIMS_FIELD);
+    }
+    functions.logger.info(
+      `Claims set for user '${uid}', logging sync time to Firestore`,
       { uid }
     );
-    return;
-  }
-  await auth.setCustomUserClaims(uid, data);
-
-  const fpath = ["_synced"];
-  if (CLAIMS_FIELD) {
-    fpath.unshift(CLAIMS_FIELD);
-  }
-  functions.logger.info(
-    `Claims set for user '${uid}', logging sync time to Firestore`,
-    { uid }
-  );
-  return change.after.ref.update(
-    new admin.firestore.FieldPath(...fpath),
-    admin.firestore.FieldValue.serverTimestamp()
-  );
-});
+    return change.after.ref.update(
+      new admin.firestore.FieldPath(...fpath),
+      admin.firestore.FieldValue.serverTimestamp()
+    );
+  });

--- a/package.json
+++ b/package.json
@@ -8,6 +8,7 @@
     "lint": "prettier --list-different \"**/*.{js,md,yml,ts,json,yaml}\"",
     "clean": "lerna run --parallel clean && lerna clean",
     "build": "lerna run --parallel build",
+    "local:emulator": "cd _emulator && firebase emulators:start -P demo-test",
     "generate-readmes": "lerna run --parallel generate-readme",
     "postinstall": "lerna bootstrap --no-ci && lerna run --parallel clean && npm run build",
     "prepare": "husky install"

